### PR TITLE
Introduce option to use local payload instead of builder 

### LINKED ISSUE #6327: Support block value in `engineGetPayload`
https://github.com/ethereum/execution-apis/issues/307

We could start comparing locally produced payloads with bids coming from builder endpoint and prefer local one if it has higher or similar value of the received bid. We could add percentage diff as well, so if let's say bid isn't at least 50% more valuable than local we fallback to local.

### PR DESCRIPTION
**Pull request by zilm13**
## PR Description
Introduces option to use local payload instead of builder if local matches X% of builder's value. X is configured:
- I have some doubts I did it in appropriate options class
- I'm not sure I should unhide it right from the beginning, plus if I do it, I will need to add docs and changelog entry
- I don't like I couldn't pass default value to tests

## Fixed Issue(s)

Fixes #6327 
## Documentation

- [ ] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [ ] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

File: ethereum/executionlayer/src/test/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionLayerManagerImplTest.java [hunk 1]
```diff
@@ -234,9 +235,46 @@ public void builderGetHeaderGetPayload_shouldReturnHeaderAndPayloadViaEngineWhen
...
+        dataStructureUtil.randomPayloadExecutionContext(false, true);
+    final UInt64 slot = executionPayloadContext.getForkChoiceState().getHeadBlockSlot();
+    final BeaconState state = dataStructureUtil.randomBeaconState(slot);
+
+    final UInt256 builderValue =
+        prepareBuilderGetHeaderResponse(executionPayloadContext, false).getValue();
     final ExecutionPayload localExecutionPayload =
-        prepareEngineGetPayloadResponse(executionPayloadContext, UInt256.MAX_VALUE, slot);
+        prepareEngineGetPayloadResponse(
+            executionPayloadContext, builderValue.multiply(51).divide(100), slot);
```
Comment (rolfyone) [inline ethereum/executionlayer/src/test/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionLayerManagerImplTest.java:277]: 50 would be the boundary right? could just multiply by boundary here... i guess point being that we're implementing >=, so boundary is 50 not 51, and we wouldnt be testing boundary at 51, just every other value...

Comment (zilm13) [reply ethereum/executionlayer/src/test/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionLayerManagerImplTest.java:277]: It's in percents

Comment (zilm13) [reply ethereum/executionlayer/src/test/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionLayerManagerImplTest.java:277]: I wanted to have 2 tests cases side by side one where value is a bit higher and another where it's a bit smaller. And I converted the old one to default value, but it's not passed from the same place in tests like in main, so it's not correct say it's default, which I don't like

Comment (rolfyone) [reply ethereum/executionlayer/src/test/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionLayerManagerImplTest.java:277]: ok but the boundary is 50%, so it'd be nice to also test the boundary and make sure >= doesnt turn into >

Comment (zilm13) [reply ethereum/executionlayer/src/test/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionLayerManagerImplTest.java:277]: ahhh got your comment now
the reason I didn't make it because it could change with even/odd number
Of course, I could use not random to be sure

Comment (zilm13) [reply ethereum/executionlayer/src/test/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionLayerManagerImplTest.java:277]: and we test == on old test, which is left in place

**Comment by zilm13** [comment]
I decided to add feature to use builder's bid whenever is possible not comparing with local value to the same flag with keyword "NEVER". If it totally doesn't work for us, I made it with a single commit, so it could be easily reverted

File: teku/src/main/java/tech/pegasys/teku/cli/options/ExecutionLayerOptions.java [hunk 2]
```diff
@@ -96,6 +97,19 @@ public class ExecutionLayerOptions {
   private int builderCircuitBreakerAllowedConsecutiveFaults =
       DEFAULT_BUILDER_CIRCUIT_BREAKER_ALLOWED_CONSECUTIVE_FAULTS;
 
+  @Option(
+      names = {"--validators-builder-bid-challenge-percentage"},
+      paramLabel = "<STRING>",
+      showDefaultValue = Visibility.ALWAYS,
+      description =
+          "Fallback to local payload, when builder bid is available but its value could be beat by local payload. "
```
Comment (tbenr) [inline teku/src/main/java/tech/pegasys/teku/cli/options/ExecutionLayerOptions.java]: we can simply say "Fallback to local payload whenever its value beats the builder bid."

File: teku/src/main/java/tech/pegasys/teku/cli/options/ExecutionLayerOptions.java [hunk 3]
```diff
@@ -96,6 +97,19 @@ public class ExecutionLayerOptions {
...
+              + "Value is whole, percent (e.g. 100, default value, means use local payload when it at least matches builder bid, "
+              + "80 means use local payload when its value is at least 80% of builder bid).\n"
+              + "NEVER: ignore local value, use builder's bid whenever it passed validation",
```
Comment (tbenr) [inline teku/src/main/java/tech/pegasys/teku/cli/options/ExecutionLayerOptions.java]: "passes"?

File: ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionBuilderModule.java [hunk 4]
```diff
@@ -164,9 +168,21 @@ public SafeFuture<HeaderWithFallbackData> builderGetHeader(
...
+                        .lessOrEqualThan(localPayloadValue.multiply(HUNDRED_PERCENTS))) {
+                  LOG.info(
+                      "Using local Execution Payload instead of Builder Bid as it's challenged. "
+                          + "Builder bid value: {}, local block value: {}, builderBidChallengePercentage: {}%",
+                      signedBuilderBid.getMessage().getValue().toDecimalString(),
+                      localPayloadValue.toDecimalString(),
+                      builderBidChallengePercentage);
                   return getResultFromLocalExecutionPayload(
-                      localExecutionPayload, slot, FallbackReason.LOCAL_BLOCK_VALUE_HIGHER);
+                      localExecutionPayload, slot, FallbackReason.LOCAL_BLOCK_VALUE_WIN);
```
Comment (tbenr) [inline ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionBuilderModule.java]: what about `LOCAL_BLOCK_VALUE_WON` ? similarly to `CIRCUIT_BREAKER_ENGAGED`

File: ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionBuilderModule.java [hunk 5]
```diff
@@ -164,9 +168,21 @@ public SafeFuture<HeaderWithFallbackData> builderGetHeader(
                         .map(ExecutionPayloadWithValue::getValue)
                         .orElse(UInt256.ZERO);
                 logReceivedBuilderBid(signedBuilderBid.getMessage());
-                if (signedBuilderBid.getMessage().getValue().lessOrEqualThan(localPayloadValue)) {
+                // 1 ETH is 10^18 wei, Uint256 max is more than 10^77
+                if (builderBidChallengePercentage.isPresent()
```
Comment (tbenr) [inline ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionBuilderModule.java]: what about moving this in a `isLocalPayloadValueWinning` function?

**Review by tbenr — APPROVED** [review]
LGTM with some nits. And maybe an english native guy can review user-facing strings 😉

File: ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionBuilderModule.java (hunk 4 above)
Comment (zilm13) [reply ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionBuilderModule.java]: it was the first name. don't know why I reverted it

**Comment by zilm13** [comment]
@tbenr updated up to your feedback. Let's keep it till native speaker for final proof-reading

**Comment by zilm13** [comment]
I've remembered what was wrong when we posted about this feature last time. It will be enabled only since Capella, so we'd better keep it hidden to avoid misleading. Docs etc should be added with Capella release I think. What do you think? (but we may have a lot of debts for this release)

**Comment by tbenr** [comment]
> I've remembered what was wrong when we posted about this feature last time. It will be enabled only since Capella, so we'd better keep it hidden to avoid misleading. Docs etc should be added with Capella release I think. What do you think? (but we may have a lot of debts for this release)

OH yes. very good catch. I was about to fall under the same issue. Even if when we start using the new engine api system that Loucas is working on, we will detach from the Capella fork.

File: ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionBuilderModule.java [hunk 6]
```diff
@@ -50,6 +50,7 @@
 public class ExecutionBuilderModule {
 
   private static final Logger LOG = LogManager.getLogger();
+  private static final int HUNDRED_PERCENTS = 100;
```
Comment (rolfyone) [inline ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionBuilderModule.java]: we shouldnt need to define 100%, but if we do, maybe just `HUNDRED_PERCENT`, as the `s` is confusing :)

File: ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionBuilderModule.java [hunk 7]
```diff
@@ -164,10 +168,18 @@ public SafeFuture<HeaderWithFallbackData> builderGetHeader(
                         .map(ExecutionPayloadWithValue::getValue)
                         .orElse(UInt256.ZERO);
                 logReceivedBuilderBid(signedBuilderBid.getMessage());
-                if (signedBuilderBid.getMessage().getValue().lessOrEqualThan(localPayloadValue)) {
+
+                if (isLocalPayloadValueWinning(signedBuilderBid, localPayloadValue)) {
+                  LOG.info(
+                      "Using local Execution Payload instead of Builder Bid as it's challenged. "
+                          + "Builder bid value: {}, local block value: {}, builderBidChallengePercentage: {}%",
```
Comment (rolfyone) [inline ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionBuilderModule.java]: `The local execution payload value ({}) was awarded the block over the builder payload ({}), {}% challenge configured.`

File: ethereum/executionlayer/src/test/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionLayerManagerImplTest.java (hunk 1 above)
Comment (rolfyone) [reply ethereum/executionlayer/src/test/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionLayerManagerImplTest.java:277]: ok fair :)

**Review by rolfyone — APPROVED** [review]
some tiny NIT's, but looks good :)

File: ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionBuilderModule.java (hunk 6 above)
Comment (zilm13) [reply ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionBuilderModule.java]: Oh, could you explain this, please, I know you are right, but I don't understand this. Is percent always plural?
